### ISSUE TITLE
Check coordinate convention of transform.warp

### ISSUE THREAD

**Issue by ahojnnes** [scikit-image/scikit-image#1789]
For all branches, make sure that we use the convention `X, Y`. Deprecate ndimage.map_coordinates behavior by internally converting from `Y, X` to `X, Y`. Improve documentation to more clearly state the expected coordinate format.

**Comment by jni** [comment]
@ahojnnes how does this fit in with the skimage convention of using (row, column) coordinates?

**Comment by JDWarner** [comment]
@jni the literature for image transforms is so steeped in the x/y convention that attempting to rock this boat is likely to mislead.

**Comment by ahojnnes** [comment]
I agree that r, c would be better. This has historic reasons. All
transformations had x, y from the beginning. And everything assumes
that format now. It would be a huge effort to change that. And I am
not sure if that is possible without breaking existing functionality
in unexpected ways.

**Comment by ahojnnes** [comment]
@JDWarner That was an argument for x, y back in the days. Now I remember as
you mention it :-)

On Nov 19, 2015, at 07:43, Josh Warner <notifications@github.com> wrote:

@jni <https://github.com/jni> the literature for image transforms is so
steeped in the x/y convention that attempting to rock this boat is likely
to mislead.

—
Reply to this email directly or view it on GitHub
<https://github.com/scikit-image/scikit-image/issues/1789#issuecomment-158046340>
.

**Comment by jni** [comment]
Got it. (Though sad to get it. =P) At any rate, though, perhaps we should have big bold warnings about the x/y coordinate convention used here and that it differs from the package conventions (and link to those).

**Comment by ahundt** [comment]
Just wanted to mention this issue was created because the (y,x) convention for this is what tripped me up and led to the creation of this issue in the first place. 

Here is the problem I faced when adding to one of the examples:
https://github.com/scikit-image/scikit-image/issues/1749

Here is the workaround for the current code, which could be dropped when this issue is completed:
https://github.com/scikit-image/scikit-image/pull/1752

**Comment by stefanv** [comment]
This remains a tricky issue.  All transformation literature uses ``(x, y)``, but the entire rest of scikit-image is ``(row, column)``.  Should we bite the bullet and change this?  It will break all the transformation codes in existence already, and make it harder to implement papers "line for line".  But consistency is a big deal, and regardless of all the pain I see here, it may be worth doing.

**Comment by jni** [comment]
@stefanv Having never been exposed to the transforms literature, I'd favour this change. =P Having said that, it's sloggy, unglamorous work, so finding someone to do it might be a challenge...

**Comment by ahojnnes** [comment]
-1.

**Comment by stefanv** [comment]
I will defer to Johannes on this one, as he is the person maintaining most
of this code.

**Comment by ahundt** [comment]
If this isn't changed it may be wise to at least make situations where the data you have will be the reverse of the data you need very clear, perhaps via comments/documentation.

**Comment by stefanv** [comment]
@ahundt Yes, that should definitely be done.  We are tagging a new release this weekend, and doing a sprint today and tomorrow to get things in shape.  Is there any chance that you can help us out on this issue?

**Comment by ahundt** [comment]
Unfortunately I'm at the end of the semester so there are lots of deadlines next week. I could in a week or so, very sorry to disappoint!

**Comment by stefanv** [comment]
No problem, we'd be very happy to have these improvements in the next release.

**Comment by soupault** [comment]
I'd agree with @jni that we need at least temporary solution (warning on convention change).

Also, there seems to be some missalignment in API. Here is the working example:
```
tr0 = transform.estimate_transform('similarity', coords_dst, coords_src)
# Used _tr0_ for image, extra transpose
img_trans = transform.warp(img.transpose(), tr0)
img_trans = img_trans.transpose()
# Use _tr0.inverse_ for points
coords_trans = tr0.inverse(coords_src)
```
I bet that user might be not aware of required `transpose` call.

**Comment by hsiaoyi0504** [comment]
As of 0.15, do we still need to transpose first ?
It seems to me this issue hasn't been closed means that it’s still the case. Alternatively, can we just mention this in the document ?

**Comment by jni** [comment]
@hsiaoyi0504 yes, the situation here is still a mess, but it is literally my top priority to fix this for 2019. Short term, where would be the most useful place to put a warning in your opinion?

**Comment by jni** [comment]
@ahojnnes @JDWarner I've thought about this a lot more than I had when we were discussing here originally regarding xy vs rc. I defer to @zpincus's great response [here](
https://github.com/scikit-image/scikit-image/pull/3031#issuecomment-390941115), where he eloquently makes the point that the issue is an **IO/display** issue, *not* an axis transposition issue, and transformation matrix rows/columns should correspond exactly with array axes (ie following `ndimage.map_coordinates` convention).

**Comment by stefanv** [comment]
I agree, perhaps more strongly, with your and my previous assessment, that we should switch.

I also caution that this will break a lot of code in pretty bad ways, so I ask that we be extra careful in warning users, or changing the API in such a way that they don't inadvertently break their code.  The plan of going via an intermediary `transform_rc`, `transform_xy` cycle seemed good.

**Comment by jni** [comment]
@stefanv Yes, we are on the same page regarding transitioning. =) The only change I would make to that plan is to change `transform_rc` to `ndtransform` or `ndtransforms`, or any other name that could be permanent.

**Comment by hsiaoyi0504** [comment]
@jni In my opinion, have comments directly embedded in the example and [notes-on-the-order-of-array-dimensions](http://scikit-image.org/docs/dev/user_guide/numpy_images.html#notes-on-the-order-of-array-dimensions) will be good.

**Comment by stefanv** [comment]
This is an annoying one, @jni.  `transform` is a good name for that submodule; having to call it something odd just so we can make behavior consistent is frustrating.  Perhaps we can just have a warning raised for three releases or so on import of the `transform` module?  It's a bit annoying to users, but at least it only happens once, and we can add a mechanism to silence it, e.g. `from skimage.__future__ import transforms` returns the transforms module, but without any noise (and the warning can also mention this workaround).

EDIT:  ^-- Inadvertently made a typo above, which suggests another solution. `transform` (verb) to `transforms` (noun), like in `transforms`.

**Comment by jni** [comment]
@stefanv indeed, my idea is that it is annoying to have to do *two* things to silence the warnings. (1) use `transform_rc` for two releases, (2) switch back to `transform` later. We could say `transforms` but it will be annoying with tab-completion compared with `ndtransform`. I'd like a name that can live in perpetuity (or till 1.0), even if we end up overwriting `transform`.

**Comment by jni** [comment]
@hsiaoyi0504 noted! I'll be happy to review if you want to submit a suggestion, or I'll get to that ASAP. Thanks for your input!

**Comment by stefanv** [comment]
There's no reason __future__. transform cannot live in perpetuity?

**Comment by stefanv** [comment]
No, I meant future. Should maybe have been __future__, but oh well.

I don't see any other alternative, other than coming up with a new 
alternative name for the submodule. And I don't have any good ideas: every 
option so far feels clunky.

Add a keyword arg to each warp function, that can be set to 'nd' or 'xy'. 
Make it None and warn when it is not specified that xy is now being returned?

**Comment by ahojnnes** [comment]
Totally agreed that we should fix this. The proposed path of implementing the new/fixed behavior in the future module seems right. The old transform module will then just get a deprecation warning for some time.

**Comment by ahojnnes** [comment]
To clarify, by fixing I mean to change from x, y to r, c convention.
